kinds.go

```diff
@@ -0,0 +1,123 @@
+package kindsys
+
+import (
+	"context"
+	"io"
+)
+
+type KindInfo struct {
+	// Organization controlled prefix
+	Group string `json:"group"`
+
+	// Must be unique within the group
+	Kind string `json:"kind"`
+
+	// Description of the purpose of this kind
+	Description string `json:"description"`
+
+	// This kind depends on composable types
+	UsesComposition []string `json:"usesComposition,omitempty"`
+
+	// Indicate where this kind is in the dev cycle
+	Maturity Maturity `json:"maturity"`
+}
+
+type VersionInfo struct {
+	// Must be vMajor-Minor-alpha
+	// for k8s... we can't have "." and support aggregation api
+	Version string `json:"version"`
+
+	// The software version when this schema was released.
+	// NOTE(1): the version must follow semantic versioning so that the order is deterministic
+	// NOTE(2): panel plugin version is saved in dashboards.  This can be used
+	// to find the appropriate schema
+	SoftwareVersion string `json:"software"`
+
+	// Human readable descriptions of the changes in this version
+	Changelog []string `json:"changelog,omitempty"`
+
+	// The YYYY-MM-DD this version was published (or empty if not yet published)
+	Published string `json:"published,omitempty"`
+
+	// JSONSchema hash
+	Signature string `json:"signature,omitempty"` // ?? hash of the json schema
+}
+
+type MachineNames struct {
+	// This is used in k8s URLs
+	Plural string `json:"plural,omitempty"`
+
+	// Used as an alias in the display
+	Singular string `json:"singular,omitempty"`
+
+	// Optional shorter names that can be matched in a CLI
+	Short []string `json:"short,omitempty"`
+}
+
+type Kind interface {
+	// Get general information about this kind
+	GetKindInfo() KindInfo
+
+	// Get the latest version
+	CurrentVersion() string
+
+	// Get all versions
+	GetVersions() []VersionInfo
+
+	// Return a JSON schema definition for the selected version
+	// When composition is required, the slots will have an any node
+	// TODO? include an option to have `AnyOf(latest known options)`
+	GetJSONSchema(version string) (string, error)
+}
+
+// A kind that manages a k8s style resource
+type ResourceKind interface {
```

Comment:
I feel like the wording is starting to become really messy.

In the same package `kindsys`, you can read:

```
// The relationship between Resource and [Kind] is similar to the
// relationship between objects and classes in conventional object oriented
// design:
//
// - Objects are instantiated from classes. The name of the class is the type of object.
// - Resources are instantiated from kinds. The name of the kind is the type of resource.
```

which imho is super confusing in combination with this `ResourceKind` definition. So, either one or the other will require some changes.